localstack/services/sns/publisher.py

```diff
@@ -864,25 +864,87 @@ def create_unsubscribe_url(external_url, subscription_arn):
 
 
 class SubscriptionFilter:
-    def check_filter_policy_on_message_attributes(self, filter_policy, message_attributes):
-        if not filter_policy:
-            return True
+    def check_filter_policy_on_message_attributes(
+        self, filter_policy: Dict, message_attributes: Dict
+    ):
+        for criteria, conditions in filter_policy.items():
+            if not self._evaluate_filter_policy_conditions_on_attribute(
+                conditions,
+                message_attributes.get(criteria),
+                field_exists=criteria in message_attributes,
+            ):
+                return False
 
-        for criteria in filter_policy:
-            conditions = filter_policy.get(criteria)
-            attribute = message_attributes.get(criteria)
+        return True
 
-            if not self._evaluate_filter_policy_conditions(
-                conditions, attribute, message_attributes, criteria
-            ):
+    def check_filter_policy_on_message_body(self, filter_policy: dict, message_body: str):
+        try:
+            body = json.loads(message_body)
+            if not isinstance(body, dict):
                 return False
```

Comment:
Is this reachable? If json.loads succeeds it will always be a dict no? and if not the exception below is thrown.

Reply:
Yes, it is! If you `json.dumps("this is a string")`, it's actually valid! There's a test case for this also. The main `container` can also be a list. The docs specify it needs to be a valid JSON object.

Reply:
https://github.com/localstack/localstack/blob/12fa7f010757e0e8f933c58a501b3437b07d9337/tests/integration/test_sns.py#L3178-L3187

Reply:
:+1: did not consider that

Reply:
The test case is there because I got specifically bitten by it, and could not understand the reason 😄